.grit/patterns/hathora_to_new_ts_sdk.md

```diff
@@ -0,0 +1,226 @@
+function resource_type_import_source($resource) {
+  if($resource <: `RoomV1Api`) {
+    $result = `"@hathora/cloud-sdk-typescript/dist/sdk/roomv1"`
+  } else if ($resource <: `AuthV1Api`) {
+    $result = `"@hathora/cloud-sdk-typescript/dist/sdk/authv1"`
+  } else if ($resource <: `LobbyV1Api`) {
+    $result = `"@hathora/cloud-sdk-typescript/dist/sdk/lobbyv1"`
+  } else if ($resource <: `LobbyV2Api`) {
+    $result = `"@hathora/cloud-sdk-typescript/dist/sdk/lobbyv2"`
+  },
+  return $result
+}
+
+function resource_type_substitute($resource) {
+  if($resource <: `RoomV1Api`) {
+    $result = `RoomV1`
+  } else if ($resource <: `AuthV1Api`) {
+    $result = `AuthV1`
+  } else if ($resource <: `LobbyV1Api`) {
+    $result = `LobbyV1`
+  } else if ($resource <: `LobbyV2Api`) {
+    $result = `LobbyV2`
+  },
+  return $result
+}
+
+function sdk_member_name($resource) {
+  if($resource <: `RoomV1Api`) {
+    $result = `roomV1`
+  } else if ($resource <: `AuthV1Api`) {
+    $result = `authV1`
+  } else if ($resource <: `LobbyV1Api`) {
+    $result = `lobbyV1`
+  } else if ($resource <: `LobbyV2Api`) {
+    $result = `lobbyV2`
+  },
+  return $result
+}
+
+// some functions have changed names to reflect deprecation.
+// XXX not all of them, eg lobby v2 `createLocalLobby` is deprecated,
+// but not renamed.
+function deprecated_suffix($resource, $method) {
+  if ($resource <: or {`RoomV1Api`, `LobbyV1Api`}) {
+    return `Deprecated`
+  },
+  return ``
+}
+
+// TODO the rest of the unwrappings
+function wrapper_key($method) {
+  if ($method <: or {`loginAnonymous`, `loginGoogle`}) {
+    return `loginResponse`
+  } else if ($method <: `getLobbyInfo`) {
+    return `lobby`
+  } else {
+    return `connectionInfo`
+  },
+  return `connectionInfo`
+}
+
+// many request funs have a different arg order, usually shifting the request to the beginning,
+// so we need to rearrange $args
+// TODO: the rest of the functions
+function reorder_args($method, $args) {
+  if ($method <: `setLobbyState`) {
+    // TODO how do we match and use rest-of-args without so much copypasta?
+    if ($args <: [$app_id, $room_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $room_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $room_id, $req]) {
+      $use_args = join(list=[$req, $room_id, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else if ($method <: `loginGoogle`) {
+    if ($args <: [$app_id, $req, $opts]) {
+      $use_args = join(list=[$req, $app_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $req]) {
+      $use_args = join(list=[$req, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else {
+    $use_args = join(list=$args, separator=", ")
+  },
+  return $use_args
+}
+
+pattern rewrite_method_calls($resource, $callee) {
+  maybe any {
+    // TODO more elegant handling of await-or-not, maybe via regex? better idea?
```

Comment:
A bit of duplication is fine. You can always pull the commonalities into a shared predicate if needed.